pkg/storage/cacher.go

```diff
@@ -355,9 +357,24 @@ func (c *Cacher) Codec() runtime.Codec {
 func (c *Cacher) processEvent(event watchCacheEvent) {
 	c.Lock()
 	defer c.Unlock()
-	for _, watcher := range c.watchers {
-		watcher.add(event)
-	}
+
+	// prevent creating too many goroutines
+	burstChan := make(chan struct{}, burstWatchers)
+	// prevent sending events to a watcher out of order
+	wait := sync.WaitGroup{}
+
+	wait.Add(len(c.watchers))
+	for _, w := range c.watchers {
+		burstChan <- struct{}{}
+		go func(watcher *cacheWatcher) {
+			defer func() {
+				<-burstChan
+				wait.Done()
+			}()
+			watcher.add(event)
+		}(w)
+	}
+	wait.Wait()
```

Comment:
This change look correct, although I'm slightly afraid about its performance impact in a healthy situation. I mean that in a healthy situation when all watchers, we are potentially adding significant overhead, because of those synchronization operations.

Do you have any benchmarks that show the impact of this change?